### DISCUSSION TITLE
prb of infinit hardening modulus

### DISCUSSION THREAD

**Discussion by GhiathMonnet** [thelfer/tfel#385] [General]
Dear All,
I have a flow rule that can be easily descibed using the @DSL IsotropicPlasticMisesFlow  by:
@FlowRule {
  if(p > 0) {
    const auto exp_Cp = exp( - C * p) ;
    const auto sig_HP = fac_HP * (1. - exp_Cp) ;
    const auto dsig_HP_dp = fac_HP * C * exp_Cp ;
    const auto sig_gamma = sig_HP + mu / 55.0 * p ;
    const auto dsig_gamma_dp = dsig_HP_dp + mu / 55.0 ;

    const auto exp_3Yp = exp(-3.* Y * p);
    const auto R = sqrt(1. - exp_3Yp);
    const auto dR_dp = 1.5 * Y * exp_3Yp / R;
    const auto sig_alpha = sigf + fac_R * R ;
    const auto dsig_alpha_dp = fac_R * dR_dp ;
    f = seq - (delta * sig_alpha + (1.- delta) * sig_gamma) ;
    df_dseq = 1;
    df_dp = -delta * dsig_alpha_dp - (1. - delta) * dsig_gamma_dp ;
  }
}

The problem is that R is a square root that goes to zero when p goes to zero. Its derivative thus goes to infinity when at the begning of loading (at p = 0). If I use the condition if (p > 0) as shown above, the behavior remains always elastic. 
Is there a clean way to get rid off this problem ?
Many thanks.
Ghiath

**Comment by thelfer** [comment]
Hi Giath,

I would introduce a small parameter p0 (a value of 1.e-8 shall be fine) in the definition of `exp_3Yp` as follows:

```cxx
const auto exp_3Yp = exp(-3.* Y * (p+p0));
```

**Reply by GhiathMonnet** [reply]
Many thanks, Thomas. Simple and efficient solution.

**Comment by pbouda95** [comment]
Hello @GhiathMonnet ,

My view is you can introduce a change of variable such as `p = ln(1+z^2)` and update your set of equations.
A quick look on your square root equation tells me that all your problems should vanish ...

Cheers,

**Reply by GhiathMonnet** [reply]
Hi @pbouda95.
Many thanks for the help. I will try his solution too.
Cheers.
Ghiah

**Reply by GhiathMonnet** [reply]
correction "this" instead of "his"
